docs/docs/cmd/login.md

```diff
@@ -23,7 +23,10 @@ m365 login [options]
 : Password for the user. Required when `authType` is set to `password`
 
 `-c, --certificateFile [certificateFile]`
-: Path to the file with certificate private key. Required when `authType` is set to `certificate`
+: Path to the file with certificate private key. Required when `authType` is set to `certificate` and `certificateBase64Encoded` option is not specified
+
+`--certificateBase64Encoded [certificateBase64Encoded]`
```

Comment:
Does it have to be a .pfx? Could we also allow using base64-encoded .pem here?

Reply:
I don't mention .pfx anywhere? You need to provide a certificate that contains a private key, this can be a .pem or a .pfx as long as it is one file containing both the certificate and the private key.
That statement is true for --certificateFile and for --certificateBase64Encoded.

Reply:
Sorry, I added the comment to the wrong line. In the line below you say:

> Base64 encoded string with a certificate private key.

which to me equals .pfx or is it possible to have a .pem with a private key as well?

Reply:
Aha, it is possible and it even is the ONLY way you can use a .pem file for authentication to Office 365 as the Private Key is vital to make the connection. A .pem file doesn't HAVE to contain a private key, but it is required to in this scenario.
(And it isn't a change with regards to how it is currently implemented, as a I said: also with --certificateFile it has to be a .pem with private key)

The .pem file would look like this (added ... for brevity and to protect my certificate):
```
Bag Attributes
    localKeyID: 5C C6 DC 96 A9 F5 38 25 64 9A 29 21 26 91 60 03 03 47 71 F1 
subject=C = AU, ST = Some-State, O = Internet Widgits Pty Ltd

issuer=C = AU, ST = Some-State, O = Internet Widgits Pty Ltd

-----BEGIN CERTIFICATE-----
MIIDazCCAlOgAwIBAgI...2fxf+T6+/BXwv/zNT9GanHdBtNsZeMz6m2q8xWcWRT
upkNnyxFLcxyK1QJXL0bp2lgwftcOk5gwQWrZUrMrBmWNKjCt151b9MXcqTgCx3R
POZc40wYquEvap5k90oy
-----END CERTIFICATE-----
Bag Attributes
    localKeyID: 5C C6 DC 96 A9 F5 38 25 64 9A 29 21 26 91 60 03 03 47 71 F1 
Key Attributes: <No Attributes>
-----BEGIN PRIVATE KEY-----
MIIEvgIBADAN...KBgD2UNn4EuaGLmy9c2+Mr
+FqVb4a6xPTAY9a6mKb1BRKpGJsx/JYHgkJAwSOQNJRBvucO9/CqVL5IcJJq2bFU
V+1KyLlPKVfrSvZLOXTz1DeSHpduVKZ6L8vj3sTAH1Vp7tIykTvMPRootScHb9v0
I6ejJgWvbEFagzEbiQ2hV0Jo
-----END PRIVATE KEY-----
```

It is also the only supported format for PEM in the current code base (nothing changes here with my PR), having one CERTIFICATE section and one PRIVATE KEY section.
PEM could have other markers but there is no code to handle them, nor have I any clue if a PEM certificate in any of those formats would work for Office 365 authentication.
Markers taken from [openssl.pem.h](https://git.openssl.org/?p=openssl.git;a=blob;f=include/openssl/pem.h;hb=HEAD)
```
  32 # define PEM_STRING_X509_OLD     "X509 CERTIFICATE"
  33 # define PEM_STRING_X509         "CERTIFICATE"
  34 # define PEM_STRING_X509_TRUSTED "TRUSTED CERTIFICATE"
  35 # define PEM_STRING_X509_REQ_OLD "NEW CERTIFICATE REQUEST"
  36 # define PEM_STRING_X509_REQ     "CERTIFICATE REQUEST"
  37 # define PEM_STRING_X509_CRL     "X509 CRL"
  38 # define PEM_STRING_EVP_PKEY     "ANY PRIVATE KEY"
  39 # define PEM_STRING_PUBLIC       "PUBLIC KEY"
  40 # define PEM_STRING_RSA          "RSA PRIVATE KEY"
  41 # define PEM_STRING_RSA_PUBLIC   "RSA PUBLIC KEY"
  42 # define PEM_STRING_DSA          "DSA PRIVATE KEY"
  43 # define PEM_STRING_DSA_PUBLIC   "DSA PUBLIC KEY"
  44 # define PEM_STRING_PKCS7        "PKCS7"
  45 # define PEM_STRING_PKCS7_SIGNED "PKCS #7 SIGNED DATA"
  46 # define PEM_STRING_PKCS8        "ENCRYPTED PRIVATE KEY"
  47 # define PEM_STRING_PKCS8INF     "PRIVATE KEY"
  48 # define PEM_STRING_DHPARAMS     "DH PARAMETERS"
  49 # define PEM_STRING_DHXPARAMS    "X9.42 DH PARAMETERS"
  50 # define PEM_STRING_SSL_SESSION  "SSL SESSION PARAMETERS"
  51 # define PEM_STRING_DSAPARAMS    "DSA PARAMETERS"
  52 # define PEM_STRING_ECDSA_PUBLIC "ECDSA PUBLIC KEY"
  53 # define PEM_STRING_ECPARAMETERS "EC PARAMETERS"
  54 # define PEM_STRING_ECPRIVATEKEY "EC PRIVATE KEY"
  55 # define PEM_STRING_PARAMETERS   "PARAMETERS"
  56 # define PEM_STRING_CMS          "CMS"
```
Same is true for .pem files containing a certificate chain (and thus multiple CERTIFICATE sections): no support in the code base, no idea if they can be used for O365 authentication.

Reply:
Thanks for the clarification. So it comes down to it that .pem and .pfx are just different formats and whether they do or don't have a private key is irrelevant to the format.

Reply:
That's a good way to summarize my long explanation 🤣

Reply:
Sometimes you need the long explanation first to get the tl;dr 😄